apps/emqx_durable_storage/src/emqx_ds_replication_layer_egress.erl

```diff
@@ -151,80 +154,193 @@ terminate(_Reason, _S) ->
 %% Internal functions
 %%================================================================================
 
+enqueue(
+    From,
+    Sync,
+    Atomic,
+    Msgs,
+    BatchSize,
+    BatchBytes,
+    S0 = #s{n = NMsgs0, n_bytes = NBytes0, queue = Q0, pending_replies = Replies0}
+) ->
+    %% At this point we don't split the batches, even when they aren't
+    %% atomic. It wouldn't win us anything in terms of memory, and
+    %% EMQX currently feeds data to DS in very small batches, so
+    %% granularity should be fine enough.
+    NMax = application:get_env(emqx_durable_storage, egress_batch_size, 1000),
+    NBytesMax = application:get_env(emqx_durable_storage, egress_batch_bytes, infinity),
+    NMsgs = NMsgs0 + BatchSize,
+    NBytes = NBytes0 + BatchBytes,
+    case (NMsgs >= NMax orelse NBytes >= NBytesMax) andalso (NMsgs0 > 0) of
+        true ->
+            %% Adding this batch would cause buffer to overflow. Flush
+            %% it now, and retry:
+            cancel_timer(S0),
+            S1 = flush(S0),
+            enqueue(From, Sync, Atomic, Msgs, BatchSize, BatchBytes, S1);
+        false ->
+            %% The buffer is empty, we enqueue the atomic batch in its
+            %% entirety:
+            Q1 = lists:foldl(fun queue:in/2, Q0, Msgs),
+            Replies =
+                case Sync of
+                    true ->
+                        [From | Replies0];
+                    false ->
+                        gen_server:reply(From, ok),
+                        Replies0
+                end,
+            S1 = S0#s{n = NMsgs, n_bytes = NBytes, queue = Q1, pending_replies = Replies},
+            case NMsgs >= NMax orelse NBytes >= NBytes of
+                true ->
+                    cancel_timer(S1),
+                    flush(S1);
+                false ->
+                    S1
+            end
+    end.
+
 -define(COOLDOWN_MIN, 1000).
 -define(COOLDOWN_MAX, 5000).
 
-do_flush(S = #s{batch = []}) ->
-    S#s{tref = start_timer()};
+flush(S) ->
+    start_timer(do_flush(S)).
+
+do_flush(S0 = #s{n = 0}) ->
+    S0;
 do_flush(
-    S = #s{batch = Messages, pending_replies = Replies, db = DB, shard = Shard}
+    S = #s{queue = Q, pending_replies = Replies, db = DB, shard = Shard, metrics_id = Metrics}
 ) ->
-    case emqx_ds_replication_layer:ra_store_batch(DB, Shard, lists:reverse(Messages)) of
+    Messages = queue:to_list(Q),
+    T0 = erlang:monotonic_time(microsecond),
+    Result = emqx_ds_replication_layer:ra_store_batch(DB, Shard, Messages),
```

Comment:
Might be a concern for a separate PR, but a discussion here for visibility won't hurt.

Currently `ra_store_batch/3` doesn't actually follow the `emqx_ds:error(_)` contract. Partially overlooked, partially because some conditions do not actually fit "retry or report error" logic, or at least not when it comes to clients.

If we leave out non-important details, performing a state change through `ra` can basically either succeed, fail (for example, when no server are reachable), or _time out_. Timeout is essentially "no knowledge if it succeeded, failed, or will succeed / fail eventually". It could happen for example when leader is alive, persisted the log entry but currently has no quorum to commit. And since we don't know the outcome, and have no idempotency on that level, we (arguably) can neither retry nor report an error to the client.

Admittedly, we can
1. Report such conditions as errors, and let `emqx_broker` decide what to do with them. I'd argue a dedicated error class might be good to have then.
1. Make the logic on that level idempotent, however it sounds quite comprehensive.

Reply:
Good point. I limited the number of retries to 0 for now.